### ISSUE TITLE
Possible bug in matching SMARTS to atoms in molecule

### ISSUE THREAD

**Issue by bannanc** [open-forcefield-group/smarty#64]
In the WIP pull requests, we have successfully made SMARTS complicated enough to find H3, however, it appears that there may be a bug in the matching of the SMARTS to reference molecules. 

For example, when testing the new methods, we created an atomtype with SMARTS: `[#1$(*-[#6](~[#8])(~[#8])-[#8])]`
which is identified as H3 in the substrucDepict ipython notebook [here](https://github.com/open-forcefield-group/open-forcefield-data/blob/master/Model-Systems/SMIRFF_prototype/smartsDepict.ipynb)

I'm attaching a screenshot of the found atom, but it's a little difficult to see when the image is this small. 
The only H3 in the molecule is identified with purple.  
![screen shot 2016-07-13 at 11 47 54](https://cloud.githubusercontent.com/assets/11968541/16815705/c3172b5c-48ef-11e6-8ebb-e14c70057d1c.png)

However, in the output, it claims that this SMARTS string is `unused in dataset; rejecting.`

**Comment by davidlmobley** [comment]
@bannanc - relevant code seems to be here and looks OK: https://github.com/open-forcefield-group/smarty/blob/master/smarty/atomtyper.py#L121

Maybe the issue is downstream with what happens to the StringData attached to the match.

Also, doublecheck that this molecule is in the AlkEthOH set?

**Comment by davidlmobley** [comment]
Wait a sec, I think I know...

**Comment by davidlmobley** [comment]
No, I don't know. It's worth noting though that at this point we're not running this on ALL of AlkEthOH, @bannanc , just a filtered subset of it (42 molecules0. It appears there is only one H3 in the set. You should probably check and make sure your SMARTS matches THAT H3.

It looks like the H3 in our set occurs in AlkEthOH_r6,

**Comment by davidlmobley** [comment]
Hmm, I actually confirmed that your pattern matches in this molecule. I wonder if explicit hydrogens get suppressed somewhere?

Let me know if you have trouble sorting this out. But definitely it's an issue with the matching/scoring not with the SMARTS.

**Comment by bannanc** [comment]
The SMARTS string should match any H3, I think. 

Also, there are 3 H3s in the current AlkEthOH set, if you start with the "answers" as initial atomtypes you'll find all 3 of them.  
Where "answers" = a set of SMARTS that could describe all of the Parm@Frost AlkEthOH atom types

**Comment by bannanc** [comment]
Oh, follow-up that might mean we need to do more investigation into the problem. 

I used the list below as the initial atomtypes:

```
 [$([#1]-C)] HC
 [#1$(*-[#6]-[#8])]    H1
 [#1$(*-[#6](~[#8])-[#8])]    H2
 [#1$(*-[#6](~[#8])(~[#8])-[#8])]    H3
 [#1$(*-[#8])]  HO
 [#6X4]  CT
 [#8X2]  OS
 [OX2+0$(*-[#1])]   OH
```

In this situation, it correctly identifies the 'H3' to all 3 H3s. I will look more carefully at the original log file tomorrow and try to identify in which step the problem is happening, if there is a problem...

**Comment by davidlmobley** [comment]
Ah, that's a good point. We DO verify that this SMARTS string correctly matches H3 if we start with it. So, somehow something is going wrong in the evaluation of a new proposed SMARTS string which introduces H3. Odd.

**Comment by bannanc** [comment]
This also used the example pointed out above, just incase the ~ bonds made a difference, which it shouldn't.

**Comment by bannanc** [comment]
I realized since we had discussed this in the pull request I should put it here. 

This is the message from the log file

```
Attempting to create new subtype:  -> '[#1$(*-[#6](~[#8])(~[#8])-[#8])]' (c_hydrogen simply c_carbonsimplyc_oxygen (any c_oxygen) (any c_oxygen))
Atom type '[#1$(*-[#6](~[#8])(~[#8])-[#8])]' (c_hydrogen simply c_carbonsimplyc_oxygen (any c_oxygen) (any c_oxygen)) unused in dataset; rejecting.
```

If you start with the SMARTS string it recognizes it is H3, but if it's found in the middle of a run, it is unused in the data set.

**Comment by davidlmobley** [comment]
Resolved by #59
